go/pkg/sysdb/coordinator/table_catalog.go

```diff
@@ -403,47 +408,98 @@ func (tc *Catalog) CreateSegment(ctx context.Context, createSegment *model.Creat
-		if metadata != nil {
-			dbSegmentMetadataList := convertSegmentMetadataToDB(createSegment.ID.String(), metadata)
-			if len(dbSegmentMetadataList) != 0 {
-				err = tc.metaDomain.SegmentMetadataDb(txCtx).Insert(dbSegmentMetadataList)
-				if err != nil {
-					log.Error("error inserting segment metadata", zap.Error(err))
-					return err
-				}
+		var err error
+		result, err = tc.createSegmentImpl(txCtx, createSegment, ts)
+		return err
+	})
+	if err != nil {
+		log.Error("error creating segment", zap.Error(err))
+		return nil, err
+	}
+	log.Info("segment created", zap.Any("segment", result))
+	return result, nil
+}
+
+func (tc *Catalog) createSegmentImpl(txCtx context.Context, createSegment *model.CreateSegment, ts types.Timestamp) (*model.Segment, error) {
+	var result *model.Segment
+
+	// insert segment
+	collectionString := createSegment.CollectionID.String()
+	dbSegment := &dbmodel.Segment{
+		ID:           createSegment.ID.String(),
+		CollectionID: &collectionString,
+		Type:         createSegment.Type,
+		Scope:        createSegment.Scope,
+		Ts:           ts,
+	}
+	err := tc.metaDomain.SegmentDb(txCtx).Insert(dbSegment)
+	if err != nil {
+		log.Error("error inserting segment", zap.Error(err))
+		return nil, err
+	}
+	// insert segment metadata
+	metadata := createSegment.Metadata
+	if metadata != nil {
+		dbSegmentMetadataList := convertSegmentMetadataToDB(createSegment.ID.String(), metadata)
+		if len(dbSegmentMetadataList) != 0 {
+			err = tc.metaDomain.SegmentMetadataDb(txCtx).Insert(dbSegmentMetadataList)
+			if err != nil {
+				log.Error("error inserting segment metadata", zap.Error(err))
+				return nil, err
 			}
 		}
-		// get segment
-		segmentList, err := tc.metaDomain.SegmentDb(txCtx).GetSegments(createSegment.ID, nil, nil, createSegment.CollectionID)
+	}
+	// get segment
+	segmentList, err := tc.metaDomain.SegmentDb(txCtx).GetSegments(createSegment.ID, nil, nil, createSegment.CollectionID)
+	if err != nil {
+		log.Error("error getting segment", zap.Error(err))
+		return nil, err
+	}
+	result = convertSegmentToModel(segmentList)[0]
+
+	return result, nil
+}
+
+func (tc *Catalog) CreateCollectionAndSegments(ctx context.Context, createCollection *model.CreateCollection, createSegments []*model.CreateSegment, ts types.Timestamp) (*model.Collection, bool, error) {
+	var resultCollection *model.Collection
+	created := false
+
+	err := tc.txImpl.Transaction(ctx, func(txCtx context.Context) error {
+		// Create the collection using the refactored helper
+		var err error
+		resultCollection, created, err = tc.createCollectionImpl(txCtx, createCollection, ts)
 		if err != nil {
-			log.Error("error getting segment", zap.Error(err))
+			log.Error("error creating collection", zap.Error(err))
 			return err
 		}
-		result = convertSegmentToModel(segmentList)[0]
+
+		// If collection already exists, then do not create segments.
+		// TODO: Should we check to see if segments does not exist? and create them?
```

Comment:
Out of curiosity, what's holding us from implementing this TODO in this PR? Is it a lot of work?